src/vs/workbench/contrib/notebook/browser/view/renderers/webviewPreloads.ts

```diff
@@ -168,27 +168,54 @@ async function webviewPreloads(ctx: PreloadContext) {
 		readonly workspace: { readonly isTrusted: boolean };
 	}
 
-	interface ScriptModule {
-		activate(ctx?: RendererContext): Promise<RendererApi | undefined | any> | RendererApi | undefined | any;
+	interface RendererModule {
+		activate(ctx: RendererContext): Promise<RendererApi | undefined | any> | RendererApi | undefined | any;
+	}
+
+	interface KernelPreloadContext {
+		getState(): { [key: string]: unknown; };
+		setState(data: { [key: string]: unknown; }): void;
+		postMessage(msg: unknown): void;
+
+		readonly onDidReceiveKernelMessage: Event<unknown>;
+		postKernelMessage(data: unknown): void;
+	}
+
+	interface KernelPreloadModule {
+		activate(ctx: KernelPreloadContext): Promise<void> | void;
+	}
+
+	function createKernelContext(): KernelPreloadContext {
+		const vscodeApi = acquireVsCodeApi();
+		return {
+			getState: vscodeApi.getState,
+			setState: vscodeApi.setState,
+			postMessage: vscodeApi.postMessage,
+
+			onDidReceiveKernelMessage: onDidReceiveKernelMessage.event,
+			postKernelMessage: (data: unknown) => postNotebookMessage('customKernelMessage', { message: data }),
+		};
 	}
 
 	const invokeSourceWithGlobals = (functionSrc: string, globals: { [name: string]: unknown }) => {
 		const args = Object.entries(globals);
 		return new Function(...args.map(([k]) => k), functionSrc)(...args.map(([, v]) => v));
 	};
 
-	const runPreload = async (url: string, originalUri: string): Promise<ScriptModule> => {
+	const runKernelPreload = async (url: string, originalUri: string): Promise<void> => {
 		const text = await loadScriptSource(url, originalUri);
-		return {
-			activate: () => {
-				try {
-					return invokeSourceWithGlobals(text, { ...kernelPreloadGlobals, scriptUrl: url });
-				} catch (e) {
-					console.error(e);
-					throw e;
-				}
+		const isModule = /\bexport\b.*\bactivate\b/.test(text);
```

Comment:
Let me know if you can think of a better check here. Unlike with the notebook renderers, we can't simply check if `acquireVsCodeApi` appears in the file because preloads are used for third party scripts such as require.js

Reply:
Should we make this explicit? Like in browsers you do [`new Worker('foo.js', {type: 'module'})`](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker). This would avoid the guessing but also means we support both variants: classic and module.

Reply:
Since we haven't finalized the kernel preload API yet, I'd prefer only supporting modules going forward

Once python and other notebook extensions on proposed APIs adopt modules for their preloads, we can remove this ugly check and all the code for running a the preload as a global

Reply:
I like modern but I wonder if that's always possible. I know that python adds require.js as preload script and I am unsure if that can be loaded as ESM module or if that a hassle to do. In the end, the renderer is exclusive to the kernel (and being restarted when switching kernels with preloads) and therefore we can be more relaxed and allow both

Reply:
Non-ES modules can _generally_ be loaded as a "module" with the only difference being that `this` in the scope of a module is undefined. That can prevent old UMD-style modules from working, but they can be re-packed.

Reply:
@connor4312  I tried this but unfortunately loading everything as modules breaks the current python preloads.

After thinking about this problem a little more, I suggest that we either:

1. Require that kernels to tell us the type of their preloads: global vs module. Only module preloads would get access to the vscode apis. For "global" preloads, we just load the script in the notebook document 

2. Require that all preloads use modules. To support global scripts such as `require`, python would register a global module preload that then injects `require.js`  into the notebook document (potentially just by inserting a new `<script>` tag)

Reply:
I would prefer (2). Modules are the way JavaScript is going. As a propose API we don't currently have a requirement to support non-module preloads, so I would rather say that everything should be a module and extension authors can inject or otherwise repack scripts as necessary to account for that.